### PR TITLE
Fix a race in Job.join that sporadically results in normal completion

### LINKED ISSUE #1123: No CancellationException thrown when join on a crashed Job
The `job2` will throw an ArithmaticException thus making the parent job cancelled. Normally we will get a CancellationException when calling the `job2.join()` but there is otherwise.

```kotlin
suspend fun main() = runBlocking {
    log(1)
    val job = launch(Dispatchers.Unconfined) {
        log(2)
        val job2 = launch(Dispatchers.Default) {
            log(3)
            val x = 1 / 0
            log(4)
        }
        job2.join()
        log(5)
    }
    log(6)
    job.join()
    log(7)
}

val dateFormat = SimpleDateFormat("HH:mm:ss:SSS")

val now = {
    dateFormat.format(Date(System.currentTimeMillis()))
}

fun log(msg: Any?) = println("${now()} [${Thread.currentThread().name}] $msg")
```

The result may be: 

```
09:35:15:065 [main @coroutine#1] 1
09:35:15:074 [main @coroutine#2] 2
09:35:15:082 [DefaultDispatcher-worker-1 @coroutine#3] 3
09:35:15:104 [main @coroutine#2] 5
Exception in thread "main" java.lang.ArithmeticException: / by zero
	at com.bennyhuo.coroutines.sample2.exceptions.E2Kt$main$2$job$1$job2$1.invokeSuspend(e2.kt:12)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:238)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:742)
09:35:15:105 [main @coroutine#1] 6
```

When falling into the slow-path, in other words, the `joinSuspend` path:

```kotlin
public final override suspend fun join() {
    if (!joinInternal()) { // fast-path no wait
        coroutineContext.checkCompletion()
        return // do not suspend
    }
    return joinSuspend() // slow-path wait
}
```

it is gonna check the job result again here:

```kotlin
public suspend inline fun <T> suspendCancellableCoroutine(
    crossinline block: (CancellableContinuation<T>) -> Unit
): T =
    suspendCoroutineUninterceptedOrReturn { uCont ->
        val cancellable = CancellableContinuationImpl(uCont.intercepted(), resumeMode = MODE_CANCELLABLE)
        // NOTE: Before version 1.1.0 the following invocation was inlined here, so invocation of this
        // method indicates that the code was compiled by kotlinx.coroutines < 1.1.0
        // cancellable.initCancellability()
        block(cancellable)
        cancellable.getResult() // HERE!!!!!
    }
```

If the job just completed, the result will be returned.

In our case, `job2` completed with a Exception, but the state of the job will turn into `Unit` for a `ResumeOnCompletion` is installed when calling `joinSuspend` and the `ResumeOnCompletion` will be invoked immediately if the result is ready.

```kotlin
private suspend fun joinSuspend() = suspendCancellableCoroutine<Unit> { cont ->
    // We have to invoke join() handler only on cancellation, on completion we will be resumed regularly without handlers
    cont.disposeOnCancellation(invokeOnCompletion(handler = ResumeOnCompletion(this, cont).asHandler))
}
```

### PR DESCRIPTION
**Pull request by elizarov**
The race happens in the slow-path of 'join' implementation when
parent invokes join on a child coroutines that crashes and cancels
the parent.

Fixes #1123

### CODE REVIEW

**Comment by LouisCAD** [comment]
Is this planned for 1.2.1?

File: kotlinx-coroutines-core/common/src/CancellableContinuationImpl.kt [hunk 1]
```diff
@@ -170,6 +170,18 @@ internal open class CancellableContinuationImpl<in T>(
...
         val state = this.state
         if (state is CompletedExceptionally) throw recoverStackTrace(state.cause, this)
+        // if the parent job was already cancelled, then throw the corresponding cancellation exception
+        // otherwise, there is a race is suspendCancellableCoroutine { cont -> ... } does cont.resume(...)
+        // before the block returns. This getResult would return a result as opposed to cancellation
+        // exception that should have happened if the continuation is dispatched for execution later.
+        if (resumeMode == MODE_CANCELLABLE) {
+            val job = context[Job]
+            if (job != null && !job.isActive) {
+                val cause = job.getCancellationException()
```
Comment (qwwdfsad) [inline kotlinx-coroutines-core/common/src/CancellableContinuationImpl.kt]: We should recover stacktrace here as well (and in `DispatchedContinuation.run#232`)

File: kotlinx-coroutines-core/jvm/test/JobStructuredJoinStressTest.kt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        repeat(nRepeats) {
+            assertFailsWith<TestException> {
+                runBlocking {
+                    // launch in background
+                    val job = launch(Dispatchers.Default) {
+                        throw TestException("OK") // crash
+                    }
+                    try {
+                        job.join()
+                    } catch (e: CancellationException) {
```
[the remark this answers is not in the document]
Comment (elizarov) [reply kotlinx-coroutines-core/jvm/test/JobStructuredJoinStressTest.kt]: Good catch. Refactoring leftovers. Should be `assertFailsWith<CancellationException> { job.join() }`

File: kotlinx-coroutines-core/common/src/CancellableContinuationImpl.kt (hunk 1 above)
Comment (elizarov) [reply kotlinx-coroutines-core/common/src/CancellableContinuationImpl.kt]: Yes!

**Pull request merged**